applications/RANSApplication/python_scripts/formulations/turbulence_models/k_epsilon_rans_formulation.py

```diff
@@ -0,0 +1,148 @@
+        return ""
+
+
+class KEpsilonEpsilonRansFormulation(ScalarTurbulenceModelRansFormulation):
+    def GetSolvingVariable(self):
+        return KratosRANS.TURBULENT_ENERGY_DISSIPATION_RATE
+
+    def GetElementNamePrefix(self):
+        return "RansKEpsilonEpsilon"
+
+    def GetConditionNamePrefix(self):
+        return "RansKEpsilonEpsilon"
+
+
+class KEpsilonRansFormulation(TwoEquationTurbulenceModelRansFormulation):
+    def __init__(self, model_part, settings):
+        default_settings = Kratos.Parameters(r'''
+        {
+            "formulation_name": "k_epsilon",
+            "stabilization_method": "algebraic_flux_corrected",
+            "turbulent_kinetic_energy_solver_settings": {},
+            "turbulent_energy_dissipation_rate_solver_settings": {},
+            "coupling_settings":
+            {
+                "relative_tolerance": 1e-3,
+                "absolute_tolerance": 1e-5,
+                "max_iterations": 10
+            },
+            "auxiliar_process_list": [],
+            "echo_level": 0
+        }''')
+
+        settings.ValidateAndAssignDefaults(default_settings)
+
+        super().__init__(
+            model_part,
+            settings,
+            KEpsilonKRansFormulation(model_part, settings["turbulent_kinetic_energy_solver_settings"]),
+            KEpsilonEpsilonRansFormulation(model_part, settings["turbulent_energy_dissipation_rate_solver_settings"]))
+
+    def AddVariables(self):
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.DENSITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.VELOCITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.MESH_VELOCITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.NORMAL)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.VISCOSITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.KINEMATIC_VISCOSITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(Kratos.TURBULENT_VISCOSITY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.RANS_Y_PLUS)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.TURBULENT_KINETIC_ENERGY)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.TURBULENT_KINETIC_ENERGY_RATE)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.TURBULENT_ENERGY_DISSIPATION_RATE)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.TURBULENT_ENERGY_DISSIPATION_RATE_2)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.RANS_AUXILIARY_VARIABLE_1)
+        self.GetBaseModelPart().AddNodalSolutionStepVariable(KratosRANS.RANS_AUXILIARY_VARIABLE_2)
+
+        Kratos.Logger.PrintInfo(self.__class__.__name__, "Added solution step variables.")
+
+    def AddDofs(self):
+        Kratos.VariableUtils().AddDof(KratosRANS.TURBULENT_KINETIC_ENERGY, self.GetBaseModelPart())
```

Comment:
Just for me to completely understand this, the standard v-p DOFs are added prior to this ones (i.e. at some point you do `solver.AddDofs()` and this does the standard stuff plus `formulation.AddDofs()`). Am I wrong?

Reply:
Yes exactly, this formulation only adds what is required by this formulation only. VP dofs are added when you included either `FractionalStepVelocityPressureRansFormulation` or`MonolithicVelocityPressureRansFormulation`